### PR TITLE
Fix misleading full-backward-hook warning for pre-hook-only modules

### LINKED ISSUE #189093: Misleading full-backward-hook warning when only `register_full_backward_pre_hook` is used
## Issue

`nn.Module.register_full_backward_pre_hook` emits a warning worded as a
`register_full_backward_hook` warning when a module's forward inputs do not
require gradients, even though no full backward hook is registered.

The warning is:

```text
Full backward hook is firing when gradients are computed with respect to module outputs since no inputs require gradients. See https://docs.pytorch.org/docs/main/generated/torch.nn.Module.html#torch.nn.Module.register_full_backward_hook for more details.
```

This looks like a misfire or at least a misleading warning for
`register_full_backward_pre_hook`:

- the reproducer only calls `register_full_backward_pre_hook`
- it never calls `register_full_backward_hook`
- the warning points users at `register_full_backward_hook` docs
- full backward pre-hooks only receive `grad_output`, so firing from the output
  side when no inputs require grad is expected and useful

This is noisy for training-style modules whose batch input tensors usually do
not require gradients, while the module parameters do. In that case, a
full-backward pre-hook is still useful because it can prepare module state after
output gradients are available and before parameter gradients are computed.

## Minimal Repro

```python
import torch
from torch import nn


class ToyModule(nn.Module):
    def __init__(self):
        super().__init__()
        self.proj = nn.Linear(4, 4, bias=False)
        self.register_full_backward_pre_hook(lambda module, grad_output: None)

    def forward(self, x):
        return self.proj(x).relu()


model = ToyModule()
x = torch.randn(2, 4)  # does not require grad

model(x).sum().backward()
```

Pytest repro with additional controls:

```bash
python -m pytest -q repros/test_pytorch_full_backward_pre_hook_warning.py
```

The repro checks:

- one input that does not require grad: passes with the unexpected warning
- one input that requires grad: passes
- multiple inputs with no input requiring grad: passes with the unexpected warning
- multiple inputs where only one input requires grad: passes
- multiple outputs with no input requiring grad: passes with the unexpected warning
- multiple inputs/outputs where only one input requires grad: no warning is
  emitted and the test passes

## Observed Output

Validated locally with:

```text
torch=2.12.0a0+0291f960b6.nv26.04.48445190
device=cpu
```

The pytest repro passes all hook-call assertions, but it reports warnings for
the three cases where no input requires grad:

```text
repros/test_pytorch_full_backward_pre_hook_warning.py::test_single_input_no_grad
repros/test_pytorch_full_backward_pre_hook_warning.py::test_multiple_inputs_no_grad
repros/test_pytorch_full_backward_pre_hook_warning.py::test_multiple_outputs_no_input_grad
  UserWarning: Full backward hook is firing when gradients are computed with respect to module outputs since no inputs require gradients. See https://docs.pytorch.org/docs/main/generated/torch.nn.Module.html#torch.nn.Module.register_full_backward_hook for more details.
```

The controls pass without that warning:

```text
PASSED repros/test_pytorch_full_backward_pre_hook_warning.py::test_single_input_requires_grad
PASSED repros/test_pytorch_full_backward_pre_hook_warning.py::test_multiple_inputs_partially_require_grad
PASSED repros/test_pytorch_full_backward_pre_hook_warning.py::test_multiple_inputs_outputs_partially_require_grad
```

The pytest summary is:

```text
6 passed, 3 warnings
```

The observed mismatch is that:

- the repro only registers `register_full_backward_pre_hook`
- no `register_full_backward_hook` is registered
- pytest still reports the full-backward-hook warning for the no-input-grad cases

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93 @mruberry @jbschlosser @walterddr @mikaylagawarecki

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #190685

`nn.Module.register_full_backward_pre_hook` emitted a warning worded for
`register_full_backward_hook` when a module's forward inputs do not require
grad, even when no full backward hook was registered.

Root cause: in `BackwardHook.setup_output_hook` (torch/utils/hooks.py), the
`if self.input_tensors_index is None:` branch is entered whenever no module
input requires grad. Everything in that branch except the `self.grad_outputs`
cleanup concerns full backward hooks only: the warning, the empty-`grad_input`
packing, and the user-hook loop that passes it. Full backward hooks are the
sole consumers of `grad_input`; full backward pre-hooks only receive
`grad_output` (already delivered above), so firing there is expected.

Fix: gate the warning, the `grad_input` packing, and the full-backward-hook
loop on `self.user_hooks`. Only `self.grad_outputs = None` stays unconditional
so per-backward state lifetime is unchanged whether or not a full backward
hook is registered.

Test Plan:

```
python test/nn/test_module_hooks.py TestModuleHookNN.test_pre_hook_no_requires_grad_no_warning
python test/nn/test_module_hooks.py TestModuleHookNN.test_hook_no_requires_grad
```

The new test asserts a pre-hook-only module with a non-grad input emits no
"Full backward hook is firing" warning; the existing test confirms a real
full backward hook still warns, still fires, and still raises when it returns
an invalid `grad_input`.

Fixes #189093

This PR was authored with the assistance of an AI coding assistant.

### CODE REVIEW

File: test/nn/test_module_hooks.py [hunk 1]
```diff
@@ -1482,6 +1482,21 @@ def hook(mod, grad_input, grad_output):
...
 
+    def test_pre_hook_no_requires_grad_no_warning(self):
+        # A full backward pre-hook only receives grad_output, so firing from the
+        # output side when no input requires grad is expected and must not emit
+        # the full-backward-hook warning. See
+        # https://github.com/pytorch/pytorch/issues/189093
+        mod = nn.Linear(2, 3)
+        inp = torch.rand(1, 2)
+        mod.register_full_backward_pre_hook(lambda mod, gO: None)
+        with warnings.catch_warnings(record=True) as w:
```
Comment (albanD) [inline test/nn/test_module_hooks.py]: You have self.assertWarnsRegex or something along those line to simplify this.

**Review by albanD — APPROVED** [review]
Thx

File: test/nn/test_module_hooks.py (hunk 1 above)
Comment (soulitzer) [reply test/nn/test_module_hooks.py]: Good idea, apparently there is assertNotWarn! (it would warn on any other unrelated warning though hmm)
